crates/pathfinder/src/sync/checkpoint.rs

```diff
@@ -275,26 +278,35 @@ async fn handle_header_stream(
 }
 
 async fn handle_transaction_stream(
-    transaction_stream: impl futures::Stream<Item = anyhow::Result<PeerData<TransactionBlockData>>>,
+    transaction_stream: impl Stream<Item = Result<PeerData<TransactionData>, PeerData<anyhow::Error>>>
+        + Send
+        + 'static,
     storage: Storage,
     chain_id: ChainId,
+    start: BlockNumber,
 ) -> Result<(), SyncError> {
-    transaction_stream
-        .map_err(Into::into)
-        .and_then(|x| transactions::compute_hashes(x, chain_id))
-        .and_then(|x| transactions::verify_commitment(x, storage.clone()))
-        .try_chunks(100)
-        .map_err(|e| e.1)
-        .and_then(|x| transactions::persist(storage.clone(), x))
-        .inspect_ok(|x| tracing::info!(tail=%x, "Transactions chunk synced"))
-        // Drive stream to completion.
+    Source::from_stream(transaction_stream.map_err(|e| e.map(Into::into)))
+        .spawn()
+        .pipe(transactions::CalculateHashes(chain_id), 10)
+        .pipe(transactions::VerifyCommitment, 10)
+        .try_chunks(10, 10)
+        .pipe(
+            transactions::StoreTransactions {
+                db: storage.connection()?,
+                start,
+            },
+            10,
+        )
```

Comment:
I will just note that its difficult to judge if chunking transaction data is worth it. A single block already contains a lot of transaction data, so it might not be worth the code complexity to introduce multi-blocks.

With headers it makes sense since each block is just a single db row.

Reply:
True, this was a premature optimization disguised as a wild guess.